File: valkyr-python/src/valkyr/_handlers.py
"""Handler interfaces for provider and store adapters."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ProviderValue:
    """A provider value with an optional cache TTL in whole seconds."""

    value: Any
    ttl_seconds: int | None = None


class Provider(ABC):
    """Called by Valkyr to fill a cache miss."""

    @abstractmethod
    async def get(self, namespace: str, key: str) -> Any | ProviderValue | None:
        """Return a raw value, a TTL-bearing value, or ``None`` for a miss."""


class Store(ABC):
    """Called by Valkyr before committing a durable mutation."""

    @abstractmethod
    async def set(
        self, namespace: str, key: str, value: Any, ttl_seconds: int | None = None
    ) -> None:
        """Persist a single value."""

    @abstractmethod
    async def set_many(
        self,
        namespace: str,
        entries: list[dict],
        ttl_seconds: int | None = None,
    ) -> None:
        """Persist a batch of values."""

    @abstractmethod
    async def delete(self, namespace: str, key_pattern: str | None) -> None:
        """Delete matching keys."""

    @abstractmethod
    async def move(self, source: str, destination: str) -> None:
        """Move the full source context to the full destination context."""

File: valkyr-go/handlers.go
package valkyr

import (
	"context"
	"encoding/json"
	"time"
)

// ProviderValue is a provider result with an optional whole-second cache TTL.
// A nil Value is a provider miss and does not carry a TTL.
type ProviderValue struct {
	Value any
	TTL   *time.Duration
}

type Provider interface {
	Get(context.Context, string, string) (any, error)
}
type Store interface {
	Set(context.Context, string, string, json.RawMessage, *time.Duration) error
	SetMany(context.Context, string, []Entry, *time.Duration) error
	Delete(context.Context, string, *string) error
	Move(context.Context, string, string) error
}
type Entry struct {
	Key   string
	Value json.RawMessage
}
